spacy/pipeline/entity_linker.py

```diff
@@ -117,10 +121,12 @@ def __init__(
         }
         if not isinstance(kb, KnowledgeBase):
             raise ValueError(Errors.E990.format(type=type(self.kb)))
-        kb.initialize(vocab)
+        if kb.vocab is None:
```

Comment:
Is it normal that the KB's vocab is `None`? And will we have problems if `self.vocab is not kb.vocab`?

One way we can have problems from two vocab instances is that only one will be serialized, so if the component had relied on setting things in the vocab and adding strings, that state would be lost.

Reply:
This was recently changed in PR #5872. Basically to allow for a default empty KB (`"spacy.EmptyKB.v1"`), I had to take out the `Vocab` field in the constructor. It feels like a circle: the EL needs the KB, but the KB needs the Vocab of the EL. So that's why I thought the EL constructor should be setting it. But then what should happen if you already have a predefined KB with entries that required a Vocab, before you even made the EL pipe?

Reply:
But you're right that deserialization relies on the KB being initialized with self.vocab - so that it plainly wrong and requires fixing.

Reply:
Hm, well the EL factory gets the `nlp` object, so it should be able to use `nlp.vocab`?

Sorry for not seeing this in #5872, sometimes it just takes a few iterations to see all the consequences unfold.

Reply:
Don't I know it ;-)

Reply:
What do you think of the current solution, @honnibal? Basically instead of taking a `KnowledgeBase` as argument in the `EntityLinker` constructor, we take a function that creates a KB from a `Vocab`. That ensures that the EL pipe is always in charge of the creating & serializing the vocab.

Reply:
I think passing a `KnowledgeBase` instance into the `__init__` is good, but if you want the factory could take a callable that you provide the vocab to, that does make sense.

However...The complexity/intricacy possible starts to add up?

One possibility would be to move the `KnowledgeBase` to be an attribute of the `Vocab`. That would be a bit more consistent with stuff like the `Morphology` class, and it answers the question of what you'd do if you want multiple entity linkers to access the same KB. It does mean you could only have one KB per pipeline though. With this approach, you would want some sort of configuration on the NLP object to set up the knowledge-base.

Reply:
Hm, I'm really not sure about that last option :|